src/coinsselectionalgorithm.cpp

```diff
@@ -0,0 +1,361 @@
+#include "coinsselectionalgorithm.hpp"
+#include <chrono>
+#include <cstdio>
+#include <iostream>
+#include <string>
+
+/* ---------- CCoinsSelectionAlgorithm ---------- */
+
+CCoinsSelectionAlgorithm::CCoinsSelectionAlgorithm(CoinsSelectionAlgorithmType _type,
+                                                   std::vector<std::pair<CAmount, size_t>> _amountsAndSizes,
+                                                   CAmount _targetAmount,
+                                                   CAmount _targetAmountPlusOffset,
+                                                   size_t _availableTotalSize)
+                                                   : type {_type},
+                                                     problemDimension{(int)_amountsAndSizes.size()},
+                                                     maxIndex{(int)_amountsAndSizes.size() - 1},
+                                                     amounts {PrepareAmounts(_amountsAndSizes)},
+                                                     sizes {PrepareSizes(_amountsAndSizes)},
+                                                     targetAmount {_targetAmount},
+                                                     targetAmountPlusOffset {_targetAmountPlusOffset},
+                                                     availableTotalSize {_availableTotalSize}
+{
+    tempSelection = new bool[problemDimension];    
+    optimalSelection = new bool[problemDimension];
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        tempSelection[index] = false;
+        optimalSelection[index] = false;
+    }
+
+    optimalTotalAmount = 0;
+    optimalTotalSize = 0;
+    optimalTotalSelection= 0;
+
+    isRunning = false;
+    stopRequested = false;
+    solvingThread = nullptr;
+    hasCompleted = false;
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    executionMicroseconds = 0;
+    #endif
+}
+
+CCoinsSelectionAlgorithm::~CCoinsSelectionAlgorithm()
+{
+    // solvingThread stopping must be performed before data destruction
+    StopSolving();
+    delete[] amounts;
+    delete[] sizes;
+    delete[] tempSelection;
+    delete[] optimalSelection;
```

Comment:
With vectors and/or smart pointers we can get rid of this

Reply:
An analogous consideration about `tmpSelection`